src/pretix/control/views/item.py

```diff
@@ -283,6 +284,32 @@ def get_queryset(self):
         return self.request.event.questions.prefetch_related('items')
 
 
+@event_permission_required("can_change_items")
+def reorder_questions(request, organizer, event):
+    try:
+        ids = json.loads(request.body.decode('utf-8'))['ids']
+    except (JSONDecodeError, KeyError):
+        return HttpResponseBadRequest("expected JSON: {ids:[]}")
+
+    questions = request.event.questions.filter(id__in=ids)
+
+    if questions.count() != len(ids):
+        raise Http404(_("Some of the provided question ids are invalid."))
+
+    positions = questions.values_list('position', flat=True)
+
+    if positions.last() - positions.first() + 1 != questions.count():
+        return HttpResponseBadRequest("ids have to be from a consecutive range")
```

Comment:
Hmm, I'm not sure that's going to work, ``position`` values can have gaps for very valid reasons. Can we instead check something like:
```
if request.event.questions.filter(position__gte=positions.first(), position__lte=positions.last()).exclude(id__in=ids).exists():
```

Reply:
Yep, just found out that `positions` are neither unique nor consecutive. Can you explain what valid reasons there are for this?
I was just working on a migration that would give all questions a consecutive position. If we don't have this, we run into a problem when more questions than fit on a single page have the same position. E.g. in the default case with `position = 0` when you just create more than 100 questions and then try to reorder them, your suggested query would reject the request (and I don't see any way to perform the reordering with sensible semantics).

Reply:
Well, first of all, up until now, this wasn't a problem, so we didn't care that much.

When using pretix' interface, gaps can easily occur when you delete a question, because we don't re-calculate all positions in this case. I don't think non-uniqueness should happen through using the interface.

However, it is possible to create non-unique positions through the REST API, since we currently neither validate nor auto-assign positions there. I would very hesitant to add in validation here since it would be a pretty breaking API change.

Therefore, a migration will not fix our problem since these situations can re-occur at any time.

Even if we *would* fix both of these issues, I would be very cautious about relying on uniquness and consecutiveness of the positions, since there is a third reason how it can break: Race conditions when deleting/sorting/adding questions at the same time. I feels like over-engineering to go through all the trouble of preventing these issues just for this feature.

> If we don't have this, we run into a problem when more questions than fit on a single page have the same position. E.g. in the default case with position = 0 when you just create more than 100 questions and then try to reorder them, your suggested query would reject the request (and I don't see any way to perform the reordering with sensible semantics).

Well, I don't think it's impossible. Even with non-uniqueness and non-conscutiveness in the position, the order is still well-defined as ``position ASC, id ASC``. We don't even need to loop over all the questions in the event as the current re-ordering code does.

Untested pseudocode:

```
input_questions = request.event.questions.filter(id__in=ids)
ordered_questions = sorted(input_questions, lambda k: ids.index(k.pk))
original_lowest_score = min((q.position, q.pk) for q in input_questions)
original_highest_score = max((q.position, q.pk) for q in input_questions)

highest_position_on_previous_page = request.event.questions.filter(
    Q(position__lt=original_lowest_score[0])
    | Q(Q(position=original_lowest_score[0]) & Q(pk__lt=original_lowest_score[1]))
).aggregate(m=Max('position'))['m'] or 0

for i, q in enumerate(ordered_questions):
    pos = highest_position_on_previous_page + 1 + i
    if pos != q.position:  # Save unneccessary UPDATE queries
        q.position = pos
        q.save(update_fields['position'])

questions_on_later_pages = request.event.questions.filter(
    Q(position__gt=original_highest_score[0])
    | Q(Q(position=original_highest_score[0]) & Q(pk__gt=original_highest_score[1]))
)
for i, q in enumerate(questions_on_later_pages):
    pos = highest_position_on_previous_page + len(input_questions) + 1 + i
    if pos != q.position:  # Save unneccessary UPDATE queries
        q.position = pos
        q.save(update_fields['position'])
```

Reply:
> Untested pseudocode:

I guess that should work. I didn't think enough about the ordering being defined by both `position` and `id`.

I think we should definitely document this behavior in more detail in the API docs. It was pretty counter intuitive to me (although I get your reasons now).
It might also be helpful to rename the field to something other than position (e.g. "sortkey" or something like that), since position (at least to me) has a pretty strong meaning (i.e. the first position, the second position, and so on). Renaming fields in a public API however is often not worth the pain. It's probably enough to extensively document the field.

Reply:
Yes. The correct name would probably be something like "priority".